jpa/eclipselink.jpa.test/src/org/eclipse/persistence/testing/tests/jpa/advanced/PersistenceUnitProcessorTest.java

```diff
@@ -35,42 +35,140 @@
+
+        // 463571
+
+        checkPURootCustom(
+            "zip", "/C:/Oracle/Middleware/Oracle_Home/wlserver/samples/domains/mydomain/servers/myserver/tmp/_WL_user/eclipselink-advanced-model/y986my/eclipselink-advanced-model_ejb.jar!/META-INF/persistence.xml",
+            "file:/C:/Oracle/Middleware/Oracle_Home/wlserver/samples/domains/mydomain/servers/myserver/tmp/_WL_user/eclipselink-advanced-model/y986my/eclipselink-advanced-model_ejb.jar"
+        );
+
+        checkPURootCustom(
+            "zip", "/C:/Program Files/Middleware/Oracle_Home/wlserver/samples/domains/mydomain/servers/myserver/tmp/_WL_user/eclipselink-advanced-model/y986my/eclipselink-advanced-model_ejb.jar!/META-INF/persistence.xml",
+            "file:/C:/Program Files/Middleware/Oracle_Home/wlserver/samples/domains/mydomain/servers/myserver/tmp/_WL_user/eclipselink-advanced-model/y986my/eclipselink-advanced-model_ejb.jar"
+        );
+
+        checkPURootCustom(
+            "zip", "/C:/Program.Files/Middleware/Oracle_Home/wlserver/samples/domains/mydomain/servers/myserver/tmp/_WL_user/eclipselink-advanced-model/y986my/eclipselink-advanced-model_ejb.jar!/META-INF/persistence.xml",
+            "file:/C:/Program.Files/Middleware/Oracle_Home/wlserver/samples/domains/mydomain/servers/myserver/tmp/_WL_user/eclipselink-advanced-model/y986my/eclipselink-advanced-model_ejb.jar"
+        );
+    }
+
+    public void testComputePURootURLForJarFile() throws Exception {
+        // Test cases for expected behavior.
+
+        checkPURootSimple(
+            "jar:file:/foo/bar.jar!/META-INF/persistence.xml",
+            "file:/foo/bar.jar"
+        );
+
+        checkPURootSimple(
+            "jar:file:/foo/bar.war!/WEB-INF/classes/META-INF/persistence.xml",
+            "jar:file:/foo/bar.war!/WEB-INF/classes/"
+        );
+
+        checkPURootSimple(
+            "jar:file:/foo/bar.jar!/WEB-INF/classes/META-INF/persistence.xml",
+            "file:/foo/bar.jar"
+        );
+
+        checkPURootSimple(
+            "jar:file:/foo/bar.jar!/foo/META-INF/persistence.xml",
+            "file:/foo/bar.jar"
+        );
+    }
+
+    public void testComputePURootURLForWsjarFile() throws Exception {
+        // Test cases for expected behavior.
+        // The results differ slightly from the other archive URLs!
+
+        checkPURootCustom(
+            "wsjar", "file:/foo/bar.jar!/META-INF/persistence.xml",
+            "jar:file:/foo/bar.jar!/"
+        );
+
+        checkPURootCustom(
+            "wsjar", "file:/foo/bar.war!/WEB-INF/classes/META-INF/persistence.xml",
+            "jar:file:/foo/bar.war!/WEB-INF/classes/"
+        );
+
+        checkPURootCustom(
+            "wsjar", "file:/foo/bar.jar!/WEB-INF/classes/META-INF/persistence.xml",
+            "jar:file:/foo/bar.jar!/"
+        );
+
+        checkPURootCustom(
+            "wsjar", "file:/foo/bar.jar!/foo/META-INF/persistence.xml",
+            "jar:file:/foo/bar.jar!/"
         );
```

Comment:
This is also not a valid persistence.xml location. I'm not sure what we should do for this situation (my gut instinct is that the URL should be validated before it gets here), but I still think we shouldnt be testing this like it's valid

Reply:
Same as the other test case - we make sure our behavior doesn't suddenly change for cases that don't conform to the JPA spec.

Reply:
That's the beauty of having a spec define your expected behavior: if it isnt supported, users shouldnt be relying on it. We don't have to make sure old, bad behavior is still supported. Instead we should throw some exception or return null or something more than providing what is otherwise a valid result.

Maybe @lukasj has a different opinion? Id be interested in thoughts on this.